New CLI subcommand to create clang-compatible compilation database (`compile_commands.json`)

qubidt opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

<!--- Describe your changes in detail here. -->

Updating @xton's PR #10264, which adds a CLI command to generate a [Clang Compilation Database file](https://clang.llvm.org/docs/JSONCompilationDatabase.html), for use with language servers, linters, and other tooling. No major changes other than addressing merge conflicts.

I changed the subcommand name, as was requested in the review for the previous PR, but to `generate-compilation-database` instead of `generate-compile-commands` as @skullydazed suggested. This seems more aligned with how the file is referred to in compatible tools' documentation. But I can change this if needed.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Fixes #4855

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

tzarc commented:
Any time I try to execute it gives this error, regardless of the parameters:
```☒ Could not determine keyboard!
☒ You must supply both `--keyboard` and `--keymap`, or be in a directory for a keyboard or keymap.```

qubidt commented:
> Any time I try to execute it gives this error, regardless of the parameters

Hmm, I have those settings saved to my config so I didn't notice that. I'll figure it out

> You'll also need to remove the debugging statements when printing out loading of modules

agh I didn't realize I committed those print statements. that's embarrassing 😅

qubidt commented:
Fixed the issue. I missed the config parameters when I renamed the subcommand, so they weren't being parsed correctly.

Also added error handling for the case where no commands were generated (e.g. you specified a non-existent keymap) and some minor improvements (instead of checking checking exit status or calling `which ...` ourselves, delegate it to `Popen`/`shutil`, respectively)

skullydazed commented on lib/python/qmk/cli/generate/compilation_database.py:
```diff
@@ -0,0 +1,125 @@
+        subprocess.run(clean_command, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, env=env, check=True)
+
+        cli.log.info('Gathering build instructions from {fg_cyan}%s', ' '.join(command))
+        with subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, env=env) as proc:
```
I don't have time to do a full review right now, but please use `cli.run` here as it handles running commands on windows correctly. Using subprocess directly in a way that works cross-platform is error-prone.

qubidt replied:
Good call, I wasn't aware of milc's `cli.run`. That'll be straightforward to fix.

As for cross-platform compatibility, I wasn't thinking of that at all. I'll have to spin up a windows environment to test it, but I don't think the way @xton searched for the system library path is very portable. Not sure how I"ll tackle that yet but any suggestions are welcome.

qubidt commented:
Now using `cli.run()` instead of `subprocess.Popen()`. It now waits for the command to complete before parsing the output from `make` but there's relatively little output so the impact to overall run time is negligible.

We will still likely need to fix the `system_libs()` function to give reasonable output for windows environments. I literally never use windows so I'll have to do some research to figure out what that will look like.

AWare commented:
Just a quick note of thanks, this PR has helped me make vscode a lot happier about qmk 👏

Diaoul commented:
Trying to use this I still have a few issues in vscode in keymaps after generating the database:
```
Unknown type name 'uint16_t' (fix available) clang(unknown_typename)
```
and 
```
Expected ';' after top level declarator (fix available) clang(invalid_token_after_toplevel_declarator)
```

Do I need specific clangd options or anything I am missing? Compiling an flashing work fine though :thinking:

user78552 commented:
@Diaoul Hi, did you manage to solve this problem?

Diaoul commented:
No, I gave up.
